### PR TITLE
[FIX] project: limit task recursion until date

### PR DESCRIPTION
**Pull request by Yoann-bary**
Steps to reproduce:
- Project > New Task > Set 'Deadline' to today
- 'Repeat Every' 1 week Until tomorrow
- Save > Change task state to 'Done'

A recurring task is created with a deadline beyond the limit date, this can be repeated indefinitely. This happens because we check he limit date against today instead of comparing it to the task deadline. This is most likely an artifact of 85e9290711c5376660941122dffb3b335b223091, where we allowed immediate recurring task creation (Which would have previously been handled by the CRON at the appropriate date).

opw-4210251

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/project/models/project_task_recurrence.py [hunk 1]
```diff
@@ -81,13 +81,15 @@ def _get_recurrence_delta(self):
 
     def _create_next_occurrence(self, occurrence_from):
         self.ensure_one()
-        if self.repeat_type == 'until' and fields.Date.today() > self.repeat_until:
-            return
```
Comment (nle-odoo) [inline addons/project/models/project_task_recurrence.py:85]: if we remove this, doesn't it mean that if we could create a recurrence in the past before today? when before we would not create anything it seems (but possibly it would still not create it if we only create the next meeting after the previous one (since the previous one should be in the past))

**Review by nle-odoo — APPROVED** [review]
as pre-review: seems good to me
functionally: I don't know this code

File: addons/project/models/project_task_recurrence.py (hunk 1 above)
Comment (Yoann-bary) [reply addons/project/models/project_task_recurrence.py:85]: I guess keeping this check could make sense as creating tasks in the past is rather useless, but that does not seem functionally invalid to me (Maybe if they were late to deliver they would still want the recurring task to be created, idk). Let's see what the review team has to say.

File: addons/project/models/project_task_recurrence.py [hunk 2]
```diff
@@ -81,13 +81,15 @@ def _get_recurrence_delta(self):
...
         # Prevent double mail_followers creation
         self = self.with_context(mail_create_nosubscribe=True)
-        self.env['project.task'].sudo().create(
-            self._create_next_occurrence_values(occurrence_from)
+        create_values = self._create_next_occurrence_values(occurrence_from)
+        exceeds_limit_date = any(
+            create_values[field] and create_values[field].date() > self.repeat_until
+            for field in self._get_recurring_fields_to_postpone()
         )
+        if not (self.repeat_type == 'until' and exceeds_limit_date):
```
Comment (auon-odoo) [inline addons/project/models/project_task_recurrence.py:91]:
```suggestion
        if not (self.repeat_type == 'until' and create_values['date_deadline'] > self.repeat_until):
```
Isn't that enough? :slightly_smiling_face:

File: addons/project/tests/test_project_recurrence.py [hunk 3]
```diff
@@ -102,23 +102,18 @@ def test_recurrence_until(self):
...
+            form.repeat_interval = 1
             form.repeat_unit = 'month'
             form.repeat_type = 'until'
-            form.repeat_until = self.date_01_01 + relativedelta(months=1)
+            form.repeat_until = self.date_01_01 + relativedelta(months=1, days=1)
+            form.date_deadline = self.date_01_01
             task = form.save()
 
-        with freeze_time(self.date_01_01 + relativedelta(days=30)):
-            task.state = '1_done'
```
Comment (auon-odoo) [inline addons/project/tests/test_project_recurrence.py:112]: Why did you delete that?

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thanks for the fix :fire:

File: addons/project/tests/test_project_recurrence.py (hunk 3 above)
Comment (Yoann-bary) [reply addons/project/tests/test_project_recurrence.py:112]: The current date doesn't  matter anymore since we're only comparing the task's date(s) to the recursion limit (Unless we go with nle's suggestion). Previously we used to compare date.today() to the limit instead.

Comment (auon-odoo) [reply addons/project/tests/test_project_recurrence.py:112]: Euh you're right :ok_hand:

File: addons/project/models/project_task_recurrence.py (hunk 2 above)
Comment (Yoann-bary) [reply addons/project/models/project_task_recurrence.py]: Done

**Review by auon-odoo — APPROVED** [review]
LGTM, thanks again!
@robodoo r+
